cranelift/codegen/src/isa/x64/inst.isle

```diff
@@ -2933,49 +2944,43 @@
         (_ Unit (emit (MInst.MulX size src1 src2 (writable_invalid_gpr) dst))))
     dst))
 
+(decl x64_mul_lo_with_flags_paired (Type bool Gpr GprMem) ProducesFlags)
+(rule (x64_mul_lo_with_flags_paired ty signed src1 src2)
+      (asm_produce_flags (x64_mul_raw ty signed src1 src2)))
+
 ;; Get the invalid register as writable
 (decl writable_invalid_gpr () WritableGpr)
 (extern constructor writable_invalid_gpr writable_invalid_gpr)
 
-;; Helper for creating `mul` instructions or `imul` instructions (depending
-;; on `signed`) for 8-bit operands.
-(decl x64_mul8 (bool Gpr GprMem) Gpr)
-(rule (x64_mul8 signed src1 src2)
-      (let ((dst WritableGpr (temp_writable_gpr))
-            (_ Unit (emit (MInst.Mul8 signed src1 src2 dst))))
-        dst))
-
 ;; Helper for creating `imul` instructions.
 (decl x64_imul (Type Gpr GprMem) Gpr)
-(rule (x64_imul ty src1 src2)
-      (let ((dst WritableGpr (temp_writable_gpr))
-            (size OperandSize (raw_operand_size_of_type ty))
-            (_ Unit (emit (MInst.IMul size src1 src2 dst))))
-        dst))
+(rule (x64_imul $I16 src1 src2) (x64_imulw_rm src1 src2))
+(rule (x64_imul $I32 src1 src2) (x64_imull_rm src1 src2))
+(rule (x64_imul $I64 src1 src2) (x64_imulq_rm src1 src2))
 
-;; Helper for creating `imul` instructions with an immediate operand.
+;; Helper for creating `imul` instructions with an immediate operand. Match
+;; 8-bit immediates first to allow a smaller instruction encoding.
 (decl x64_imul_imm (Type GprMem i32) Gpr)
-(rule (x64_imul_imm ty src1 src2)
-      (let ((dst WritableGpr (temp_writable_gpr))
-            (size OperandSize (raw_operand_size_of_type ty))
-            (_ Unit (emit (MInst.IMulImm size src1 src2 dst))))
-        dst))
+(rule 2 (x64_imul_imm $I16 src1 (is_rust_simm8 src2)) (x64_imulw_rmi_sxb src1 src2))
+(rule 2 (x64_imul_imm $I32 src1 (is_rust_simm8 src2)) (x64_imull_rmi_sxb src1 src2))
+(rule 2 (x64_imul_imm $I64 src1 (is_rust_simm8 src2)) (x64_imulq_rmi_sxb src1 src2))
+(rule 1 (x64_imul_imm $I16 src1 (is_rust_imm16 src2)) (x64_imulw_rmi src1 src2))
+(rule 1 (x64_imul_imm $I32 src1 (is_rust_imm32 src2)) (x64_imull_rmi src1 src2))
+(rule 1 (x64_imul_imm $I64 src1 src2)                 (x64_imulq_rmi_sxl src1 src2))
```

Comment:
Once again (as in #10762), we run into ISLE difficulties matching immediates. I did put up #10775, but I'm not really sure if it helps all that much. Another option to avoid these extra conversions would be to wrap up  the `i32` in a `RegMemImm.Imm` and then immediately unwrap it, but that confuses the type declaration. Not sure what to do, but it seems this is ripe for some refactoring...

Reply:
I'm a bit confused, how is this related to #10762? Here the argument is `i32`, not a `*MemImm`, so this to me at least looks like a normal what I'd expect "match the size/value of the immediate and emit an appropriate instruction"

Reply:
What I mean is that it's not easy to line up all the types so that matching can be clear and simple; this is not a problem with ISLE per se but with the network of rules that currently exist in the `*.isle` files.

Reply:
Hm I'm not sure I understand still, but nevertheless this whole PR looks quite reasonable to me so I wouldn't have any qualms about landing it myself.